### PR TITLE
MULE-12126: The SDK should communicate to the Expression Manager the …

### PR DESCRIPTION
**Pull request by estebanwasinger**
…expected DataType

### CODE REVIEW

File: core-tests/src/test/java/org/mule/runtime/core/util/DWAttributeEvaluatorTestCase.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  private static final String HOST_PORT_JSON = "{\"host\":\"0.0.0.0\", \"port\" : 8081}";
+  private static final String JSON_CAR = "{\n  \"color\": \"RED\",\n  \"price\": 1000\n}";
+  private static final String DW_CAR = "#[{color : 'RED', price: 1000}]";
+  private static final DataType CAR_DATA_TYPE = DataType.fromType(Car.class);
+  private Event mockMuleEvent = mock(Event.class);
+  private DefaultExpressionManager expressionManager;
+
+  @Before
+  public void setUp() throws RegistrationException {
+    DefaultStreamingManager streamingManager = new DefaultStreamingManager();
```
Comment (marianogonzalez) [inline core-tests/src/test/java/org/mule/runtime/core/util/DWAttributeEvaluatorTestCase.java]: you can just use a mock

File: core/src/main/java/org/mule/runtime/core/el/DefaultExpressionManager.java [hunk 2]
```diff
@@ -150,28 +142,32 @@ public TypedValue evaluate(String expression, DataType outputType, BindingContex
...
+      value = new TypedValue<>(streamingManager.manage((CursorProvider) payload, event), value.getDataType());
+    }
+
+    return value;
   }
 
   private TypedValue transform(TypedValue target, DataType sourceType, DataType outputType) throws TransformerException {
-    Object result = muleContext.getRegistry().lookupTransformer(sourceType, outputType).transform(target.getValue());
-    return new TypedValue<>(result, outputType);
+    if (!isInstance(outputType.getType(), target.getValue()) && target.getValue() != null) {
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/el/DefaultExpressionManager.java]: invert order

File: core/src/main/java/org/mule/runtime/core/util/AttributeEvaluator.java [hunk 3]
```diff
@@ -7,38 +7,93 @@
...
-  private static final Pattern SINGLE_EXPRESSION_REGEX_PATTERN = Pattern.compile("^#\\[(?:(?!#\\[).)*\\]$", DOTALL);
-
-  private enum AttributeType {
-    EXPRESSION, PARSE_EXPRESSION, STATIC_VALUE
-  }
-
-  private final String attributeValue;
+  private static final Pattern SINGLE_EXPRESSION_REGEX_PATTERN = Pattern.compile("^#\\[(?:(?!#\\[).)*]$", DOTALL);
+  private static final BindingContext NULL_BINDING_CONTEXT = BindingContext.builder().build();
+  private static final List<Class<?>> blackListTypes = Arrays.asList(Object.class, InputStream.class);
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/util/AttributeEvaluator.java]: caps

File: core/src/main/java/org/mule/runtime/core/util/AttributeEvaluator.java [hunk 4]
```diff
@@ -7,38 +7,93 @@
...
+      case PARSE_EXPRESSION:
+        configureParseAttribute();
+        break;
+      case STATIC_VALUE:
+        configureStaticAttribute();
+    }
+  }
+
+  private void configureStaticAttribute() {
+    parseResolver = (event -> this.attributeValue);
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/util/AttributeEvaluator.java]: redundant parenthesis in both lines

File: core/src/main/java/org/mule/runtime/core/util/AttributeEvaluator.java [hunk 5]
```diff
@@ -7,38 +7,93 @@
...
+    expressionResolver = (event -> new TypedValue<>(this.attributeValue, this.attributeValue == null ? OBJECT : STRING));
+  }
+
+  private void configureParseAttribute() {
+    parseResolver = (event -> expressionManager.parse(this.attributeValue, event, null));
```
Comment (marianogonzalez) [inline core/src/main/java/org/mule/runtime/core/util/AttributeEvaluator.java]: parenthesis

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSet.java [hunk 6]
```diff
@@ -83,11 +86,27 @@ public boolean isDynamic() {
...
+  /**
+   * Evaluates all the added {@link ValueResolver}s and returns the results into a {@link ResolverSetResult}
+   *
+   * @param event a not {@code null} {@link Event}
+   * @param hashable indicates if the created {@link ResolverSetResult} supports to be compared by the hash of their
+   *                 components
+   * @return a {@link ResolverSetResult}
+   * @throws MuleException if an error occurs creating the {@link ResolverSetResult}
+   */
+  public ResolverSetResult resolve(Event event, boolean hashable) throws MuleException {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSet.java]: HashedResolverSet

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/HashableResolverSetResult.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+ * resolved values. This is required for Configuration and Connection Provider cases to be able to cache the
+ * instances that are based on dynamic parameters.
+ * <p>
+ * This classes {@link #equals(Object)} and {@link #hashCode()} methods have been redefined to be consistent with the result
+ * objects. This is so that given two instances of this class you can determine if the evaluations they represent have an
+ * equivalent outcome
+ *
+ * @since 4.0
+ */
+public class HashableResolverSetResult extends ResolverSetResult {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/HashableResolverSetResult.java]: HashedResolverSetResult

File: tests/test-extensions/typed-value-extension/src/main/java/org/mule/test/typed/value/extension/extension/NullOutputResolver.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+ */
+package org.mule.test.typed.value.extension.extension;
+
+import org.mule.metadata.api.model.MetadataType;
+import org.mule.runtime.api.connection.ConnectionException;
+import org.mule.runtime.api.metadata.MetadataContext;
+import org.mule.runtime.api.metadata.MetadataResolvingException;
+import org.mule.runtime.api.metadata.resolving.OutputTypeResolver;
+
+public class NullOutputResolver implements OutputTypeResolver {
```
Comment (marianogonzalez) [inline tests/test-extensions/typed-value-extension/src/main/java/org/mule/test/typed/value/extension/extension/NullOutputResolver.java]: jdoc

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSetResult.java [hunk 9]
```diff
@@ -100,7 +100,7 @@ public Object get(String parameterName) {
    */
   @Override
   public boolean equals(Object obj) {
-    if (obj instanceof HashableResolverSetResult) {
+    if (obj instanceof HashedResolverSetResult) {
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSetResult.java]: BUUUUUGGGGGGGGG

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSetResult.java [hunk 10]
```diff
@@ -95,39 +87,30 @@ public Object get(String parameterName) {
...
+   * <p>
    * This implementation fails fast. Evaluation is finished at the first non equal value, returning {@code false}
    *
    * @param obj the object to test for equality
    * @return whether the two objects are equal
    */
   @Override
   public boolean equals(Object obj) {
-    if (obj instanceof ResolverSetResult) {
+    if (obj instanceof HashedResolverSetResult) {
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSetResult.java]: shame

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSetResult.java [hunk 11]
```diff
@@ -95,39 +87,30 @@ public Object get(String parameterName) {
...
   }
 
-  /**
-   * A hashCode calculated based on the results
-   *
-   * @return a hashCode
-   */
-  @Override
-  public int hashCode() {
-    return hashCode;
```
Comment (alegmarra) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/resolver/ResolverSetResult.java]: revert

**Pull request merged**